include/nuttx/ipcc.h

```diff
@@ -0,0 +1,212 @@
+   *   that case function can return value less then buflen.
+   *
+   * Input Parameters:
+   *   ipcc - ipcc channel to write data to
+   *   buffer - location to read data from
+   *   buflen - number of bytes to write from buffer
+   *
+   * Returned Value:
+   *   Number of actually written data to IPCC channel.
+   *
+   **************************************************************************/
+
+  CODE ssize_t (*write)(FAR struct ipcc_lower_s *ipcc,
+                        FAR const char *buffer, size_t buflen);
+
+  /**************************************************************************
+   * Name: buffer_data
+   *
+   * Description:
+   *   Lower driver shall copy data from IPCC to cirbuf on demand.
+   *   It does not have to copy all data (in case rxbuf is full), but
+   *   it must keep internal pointer to track uncopied data, and copy
+   *   what is left on the next call to this function.
+   *
+   * Input Parameters:
+   *   ipcc - channel to buffer data
+   *   rxbuf - circural buffer where data should be written to
+   *
+   * Returned Value:
+   *   Number of bytes that were successfully written to rxbuf.
+   *
+   **************************************************************************/
+
+  CODE ssize_t (*buffer_data)(FAR struct ipcc_lower_s *ipcc,
+                              FAR struct circbuf_s *rxbuf);
+
+  /**************************************************************************
+   * Name: write_notify
+   *
+   * Description:
+   *   Upper half driver notifies lower half driver that new data has been
+   *   copied to circ buffer. This can be used by lower half to transfer
+   *   data from buffer to IPCC memory.
+   *
+   * Input Parameters:
+   *   ipcc - ipcc channel that wrote data.
+   *
+   * Returned Value:
+   *   0 on success, or negated errno otherwise
+   *
+   **************************************************************************/
+
+  CODE ssize_t (*write_notify)(FAR struct ipcc_lower_s *ipcc);
+
+  /**************************************************************************
+   * Name: cleanup
+   *
+   * Description:
+   *   Cleans up resources initialized by <arch>_ipcc_init(). If arch code
+   *   malloc()ed any memory, this funciton is responsible of freeing it.
+   *   After this function is called, ipcc driver will not access lower
+   *   half pointer anymore.
+   *
+   * Input Parameters:
+   *   ipcc - ipcc channel to cleanup
+   *
+   * Returned Value:
+   *   Always OK
+   *
+   **************************************************************************/
+
+  CODE int (*cleanup)(FAR struct ipcc_lower_s *ipcc);
+};
+
+/* This structure defines the interface between upper and lower halves.
+ * Such instance is passed to ipcc_register() function after driver
+ * has been initialized, binding the upper and lower halves into one
+ * driver.
+ */
+
+struct ipcc_lower_s
+{
+  FAR struct pollfd        *fds[CONFIG_IPCC_NPOLLWAITERS];
```

Comment:
most fields in ipcc_lower_s except ops should be part of ipcc_driver_s instead

Reply:
But does it really make sens to do so? ipcc_lower_s is a shared interface between lower and upper halfs of the driver.

We have 3 structs in that driver
**ipcc_driver_s**   - accessible only by upper half of the driver
**ipcc_lower_s**    - shared struct, accessible by both upper and lower - used to communicate two parts
**stm32wl5_ipcc_s** - accessible only by lower half of the driver

So I think it makes a lot of sense to leave all these fields in lower_s, because lower driver does access them - directly or indirectly. Even pollfd is used by lower half by calling ipcc_rxfree_notify() which then calls ipcc_pollnotify() that actually uses pollfd field. If I were to move these fields to **ipcc_driver_s** then lower half of the driver would have to have acess to **ipcc_drver_s** which does not really make much sense to me as lower part does not really have to know what the driver is doing. Lower half just has to implement all interfaces and use all fields in **ipcc_lower_s** as described, and that is all that lower half cares about.

What do you think now?

Reply:
I mean that the fields which is only used by upper half.

Reply:
In that case all of these fields are used by both upper and lower half.

Buffers are directly used in lower half. fds and notify semaphores are used indirectly, that is lower half calls functions from upper half which accept ipcc_lower_s. As I understand you want these upper half functions to accept ipcc_driver_s instead? In that case I would have to add pointer to ipcc_driver_s to ipcc_lower_s, thus exposing upper driver fields to lower half.

But I guess it does make sense. And since all of these are internal APIs anyway, we could "restrict" access to upper half from lower half by comment:)

Reply:
Ok, this should be fixed now. Data that is directly used in lower half has been left in ipcc_lower_s, and data that is used (code reference) in upper half has been moved to ipcc_driver_s. ipcc_lower_s has now pointer to ipcc_driver_s structure which should only be used to call upper half functions from lower half driver.